### PR TITLE
[SPARK-5840][SQL] HiveContext cannot be serialized due to tuple extraction

### PR DESCRIPTION
**Pull request by rxin**
Also added test cases for checking the serializability of HiveContext and SQLContext.

### CODE REVIEW

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveContext.scala [hunk 1]
```diff
@@ -204,22 +204,25 @@ class HiveContext(sc: SparkContext) extends SQLContext(sc) {
...
+      state = new SessionState(new HiveConf(classOf[SessionState]))
+      SessionState.start(state)
+    }
+    state
+  }
+
+  @transient protected[hive] lazy val hiveconf: HiveConf = {
+    setConf(sessionState.getConf.getAllProperties)
+    if (sessionState.out == null) {
+      sessionState.out = new PrintStream(outputBuffer, true, "UTF-8")
```
Comment (marmbrus) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveContext.scala]: why is this here and not with the initialization of session state?

Comment (rxin) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveContext.scala]: oops

**Comment by rxin** [comment]
@marmbrus I updated it with test cases.

**Comment by marmbrus** [comment]
Merged to master and 1.3.  Thanks for adding the tests too!
